File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/executor/ImpactedTestsProvider.java
package com.teamscale.test_impacted.engine.executor;

import com.teamscale.client.ClusteredTestDetails;
import com.teamscale.client.CommitDescriptor;
import com.teamscale.client.PrioritizableTestCluster;
import com.teamscale.client.StringUtils;
import com.teamscale.client.TeamscaleClient;
import com.teamscale.test_impacted.commons.LoggerUtils;
import okhttp3.ResponseBody;
import retrofit2.Response;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Class for retrieving the impacted {@link PrioritizableTestCluster}s corresponding to {@link ClusteredTestDetails}
 * available for test execution.
 */
public class ImpactedTestsProvider {

	private static final Logger LOGGER = LoggerUtils.getLogger(ImpactedTestsProvider.class);

	private final TeamscaleClient client;

	private final String baseline;

	private final String baselineRevision;

	private final CommitDescriptor endCommit;

	private final String endRevision;

	private final String repository;

	/**
	 * The partition for the tests provided.
	 */
	public final String partition;

	private final boolean includeNonImpacted;

	private final boolean includeAddedTests;

	private final boolean includeFailedAndSkipped;

	public ImpactedTestsProvider(TeamscaleClient client, String baseline, String baselineRevision, CommitDescriptor endCommit, String endRevision, String repository, String partition,
								 boolean includeNonImpacted, boolean includeAddedTests,
								 boolean includeFailedAndSkipped) {
		this.client = client;
		this.baseline = baseline;
		this.baselineRevision = baselineRevision;
		this.endCommit = endCommit;
		this.endRevision = endRevision;
		this.repository = repository;
		this.partition = partition;
		this.includeNonImpacted = includeNonImpacted;
		this.includeAddedTests = includeAddedTests;
		this.includeFailedAndSkipped = includeFailedAndSkipped;
	}

	/** Queries Teamscale for impacted tests. */
	public List<PrioritizableTestCluster> getImpactedTestsFromTeamscale(
			List<ClusteredTestDetails> availableTestDetails) {
		try {
			LOGGER.info(() -> "Getting impacted tests...");
			Response<List<PrioritizableTestCluster>> response;
			if (!StringUtils.isBlank(endRevision)) {
				response = client
						.getImpactedTests(availableTestDetails, baseline, baselineRevision, endRevision, repository, Collections.singletonList(partition),
								includeNonImpacted, includeAddedTests, includeFailedAndSkipped);
			} else if (endCommit != null){
				response = client
						.getImpactedTests(availableTestDetails, baseline, baselineRevision, endCommit, repository,
								Collections.singletonList(partition),
								includeNonImpacted, includeAddedTests, includeFailedAndSkipped);
			} else {
				LOGGER.severe(() -> "Retrieving impacted tests failed. Please either provide an endRevision or an endCommit.");
				return null;
			}
			if (response.isSuccessful()) {
				List<PrioritizableTestCluster> testClusters = response.body();
				if (testClusters != null && testCountIsPlausible(testClusters, availableTestDetails)) {
					return testClusters;

				}
				LOGGER.severe("Teamscale was not able to determine impacted tests:\n" + response.body());
			} else {
				LOGGER.severe("Retrieval of impacted tests failed: " + response.code() + " " + response
						.message() + "\n" + getErrorBody(response));
			}
		} catch (IOException e) {
			LOGGER.log(Level.SEVERE, e, () -> "Retrieval of impacted tests failed.");
		}
		return null;
	}

	private static String getErrorBody(Response<?> response) throws IOException {
		try (ResponseBody error = response.errorBody()) {
			if (error != null) {
				return error.string();
			}
		}
		return "";
	}

	/**
	 * Checks that the number of tests returned by Teamscale matches the number of available tests when running with
	 * {@link #includeNonImpacted}.
	 */
	private boolean testCountIsPlausible(List<PrioritizableTestCluster> testClusters,
										 List<ClusteredTestDetails> availableTestDetails) {
		long returnedTests = testClusters.stream().mapToLong(g -> g.tests.size()).sum();
		if (!this.includeNonImpacted) {
			LOGGER.info(
					() -> "Received " + returnedTests + " impacted tests of " + availableTestDetails.size() + " available tests.");
			return true;
		}
		if (returnedTests == availableTestDetails.size()) {
			return true;
		} else {
			LOGGER.severe(
					() -> "Retrieved " + returnedTests + " tests from Teamscale, but expected " + availableTestDetails
							.size() + ".");
			return false;
		}
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/engine/executor/ImpactedTestsProvider.kt
package com.teamscale.test_impacted.engine.executor

import com.teamscale.client.ClusteredTestDetails
import com.teamscale.client.CommitDescriptor
import com.teamscale.client.PrioritizableTestCluster
import com.teamscale.client.TeamscaleClient
import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import retrofit2.Response
import java.io.IOException
import java.util.logging.Level

/**
 * Provides functionality to query impacted tests from Teamscale based on various configurable conditions.
 *
 * @property client The Teamscale client used to communicate with the Teamscale server.
 * @property baseline The baseline identifier for comparison, if applicable.
 * @property baselineRevision The specific baseline revision, if applicable.
 * @property endCommit The commit descriptor indicating the end of the revision range.
 * @property endRevision The specific end revision, if applicable.
 * @property repository The repository name to be used for querying impacted tests.
 * @property partition The partition key used to categorize tests for querying.
 * @property includeNonImpacted Whether to include non-impacted tests in the results.
 * @property includeAddedTests Whether to include tests that were newly added.
 * @property includeFailedAndSkipped Whether to include failed and skipped tests in the impact analysis.
 */
open class ImpactedTestsProvider(
	private val client: TeamscaleClient,
	private val baseline: String?,
	private val baselineRevision: String?,
	private val endCommit: CommitDescriptor?,
	private val endRevision: String?,
	private val repository: String?,
	val partition: String,
	private val includeNonImpacted: Boolean,
	private val includeAddedTests: Boolean,
	private val includeFailedAndSkipped: Boolean
) {
	/** Queries Teamscale for impacted tests.  */
	fun getImpactedTestsFromTeamscale(
		availableTestDetails: List<ClusteredTestDetails>
	): List<PrioritizableTestCluster>? {
		try {
			LOG.info { "Getting impacted tests..." }
			val response = client
				.getImpactedTests(
					availableTestDetails, baseline, baselineRevision, endCommit, endRevision, repository,
					listOf(partition), includeNonImpacted, includeAddedTests, includeFailedAndSkipped
				)

			if (response.isSuccessful) {
				val testClusters = response.body()
				if (testClusters != null && testCountIsPlausible(testClusters, availableTestDetails)) {
					return testClusters
				}
				LOG.severe("""
					Teamscale was not able to determine impacted tests:
					${response.body()}
					""".trimIndent())
			} else {
				LOG.severe("Retrieval of impacted tests failed: ${response.code()} ${response.message()}\n${getErrorBody(response)}")
			}
		} catch (e: IOException) {
			LOG.log(
				Level.SEVERE, e
			) { "Retrieval of impacted tests failed." }
		}
		return null
	}

	/**
	 * Checks that the number of tests returned by Teamscale matches the number of available tests when running with
	 * [includeNonImpacted].
	 */
	private fun testCountIsPlausible(
		testClusters: List<PrioritizableTestCluster>,
		availableTestDetails: List<ClusteredTestDetails>
	): Boolean {
		val returnedTests = testClusters.stream().mapToLong {
			it.tests?.size?.toLong() ?: 0
		}.sum()
		if (!includeNonImpacted) {
			LOG.info { "Received $returnedTests impacted tests of ${availableTestDetails.size} available tests." }
			return true
		}
		if (returnedTests == availableTestDetails.size.toLong()) {
			return true
		} else {
			LOG.severe {
				"Retrieved $returnedTests tests from Teamscale, but expected ${availableTestDetails.size}."
			}
			return false
		}
	}

	companion object {
		private val LOG = createLogger()

		@Throws(IOException::class)
		private fun getErrorBody(response: Response<*>): String {
			response.errorBody().use { error ->
				if (error != null) {
					return error.string()
				}
			}
			return ""
		}
	}
}
